src/meta/src/hummock/compaction/tier_compaction_picker.rs

```diff
@@ -207,21 +230,47 @@ impl TierCompactionPicker {
         select_level_handler: &LevelHandler,
         target_level_handler: &LevelHandler,
     ) -> (Vec<SstableInfo>, Vec<SstableInfo>) {
+        let mut compacting_key_range: Option<KeyRange> = None;
+        for table_info in select_level
+            .table_infos
+            .iter()
+            .filter(|t| select_level_handler.is_pending_compact(&t.id))
```

Comment:
> Only condiser compacting SSTs in L0 is not enough since SSTs in L1 may have larger key range.
> 
> ```
> Key Range:
> L0:  |____1____|            
> L1:           |__2__|
> ```
> 
> Also, we may want to check store CompactTask info (e.g. full key range) in CompactStatus so that we can do the check directly without computing the key range every time. Random thought: should we combine CompactStatus and CompactTask (i.e. make CompactTask part of CompactStatus)?

no. there may be multiple compact tasks

Reply:
> > Only condiser compacting SSTs in L0 is not enough since SSTs in L1 may have larger key range.
> > ```
> > Key Range:
> > L0:  |____1____|            
> > L1:           |__2__|
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Also, we may want to check store CompactTask info (e.g. full key range) in CompactStatus so that we can do the check directly without computing the key range every time. Random thought: should we combine CompactStatus and CompactTask (i.e. make CompactTask part of CompactStatus)?
> 
> no. there may be multiple compact tasks

I mean storing the "task id -> CompactTask" collection in CompactStatus

Reply:
> Only condiser compacting SSTs in L0 is not enough since SSTs in L1 may have larger key range.
> 
> ```
> Key Range:
> L0:  |____1____|            
> L1:           |__2__|
> ```
> 

Discussed offline. This is not an issue since we will check overlaps between L0 SST and L1 compacting SST in `pick_target_level_overlap_files` and we won't select the L0 SST if `pick_target_level_overlap_files` returns false.

Reply:
I did not extend the origin key_range of the picked table because I have check overlap in https://github.com/singularity-data/risingwave/blob/main/src/meta/src/hummock/compaction/tier_compaction_picker.rs#L241.

Reply:
`so that we can do the check directly without computing the key range every time`. 

I do not maintain the key range just for hash-compaction can use this picker by just replace the `Box<dyn OverlapStrategy>`. I think even if we build a large cluster, the number of tables locate in L1 won't be large. So the time of computing is quickly. And for level with more data, I will add another method in this trait.

Reply:
`I mean storing the "task id -> CompactTask" collection in CompactStatus`.
I suggest to maintain the LSM table infos in CompactStatus directly. Just as folloing:

```rust
struct SSTableInfo {
    ...
    pending_compact: bool
}

struct LevelInfo {
      table_infos: Vec<SSTableInfo>,
}

struct CompactStatus {
     levels: Vec<LevelInfo>,
     pending_task:  HashMap<u64, Vec<u64>>,
}

```

Reply:
> `I mean storing the "task id -> CompactTask" collection in CompactStatus`. I suggest to maintain the LSM table infos in CompactStatus directly. Just as folloing:
> 
> ```rust
> struct SSTableInfo {
>     ...
>     pending_compact: bool
> }
> 
> struct LevelInfo {
>       table_infos: Vec<SSTableInfo>,
> }
> 
> struct CompactStatus {
>      levels: Vec<LevelInfo>,
>      pending_task:  HashMap<u64, Vec<u64>>,
> }
> ```

+1